icu4c/source/test/intltest/loctest.cpp

```diff
@@ -2748,4 +2752,94 @@ void LocaleTest::TestBug13554() {
     }
 }
 
+void LocaleTest::TestMoveAssign() {
+    // ULOC_FULLNAME_CAPACITY == 157 (uloc.h)
+    Locale l1("de@collation=phonebook;x="
+              "aaaaabbbbbcccccdddddeeeeefffffggggghhhhh"
+              "aaaaabbbbbcccccdddddeeeeefffffggggghhhhh"
+              "aaaaabbbbbcccccdddddeeeeefffffggggghhhhh"
+              "aaaaabbbbbzz");
+
+    Locale l2;
+    {
+        Locale l3(l1);
+        assertTrue("l1 == l3", l1 == l3);
+        assertTrue("!l3.isBogus()", !l3.isBogus());
+        l2 = std::move(l3);
+        assertTrue("l1 == l2", l1 == l2);
+        assertTrue("l3.isBogus()", l3.isBogus());
+        assertTrue("l3.getName()", l3.getName() != nullptr);
+    }
+
+    // This should remain true also after l3 has been destructed.
+    assertTrue("l1 == l2, again", l1 == l2);
+
+    Locale l4("de@collation=phonebook");
+
+    Locale l5;
+    {
+        Locale l6(l4);
+        assertTrue("l4 == l6", l4 == l6);
+        assertTrue("!l6.isBogus()", !l6.isBogus());
+        l5 = std::move(l6);
+        assertTrue("l4 == l5", l4 == l5);
+        assertTrue("l6.isBogus()", l6.isBogus());
+        assertTrue("l6.getName()", l6.getName() != nullptr);
```

Comment:
I would also not test other behavior of the move-source objects after moving. We say that the behavior is undefined, and so we should not put specific expectations into the unit test.

An exception is that the destructor should not crash (we get that implicitly) and that there is no bad behavior, but that can be hard to test so maybe not worth it.

Optional: If you did want to test for some specifics, you could make a Locale for a really long string (to trigger heap allocation) and test that after moving the getName() pointer is different and that getVariant() does not go into the weeds. Trouble is, this might fail if we make the fullNameBuffer long enough for the string in the test. I guess we would see the test failure at that time.